cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/KVTransformation.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.batch;
+
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.InvalidEntry;
+import co.cask.cdap.etl.api.Transformation;
+import co.cask.cdap.etl.api.batch.BatchAggregator;
+import co.cask.cdap.etl.api.batch.BatchJoiner;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.common.Constants;
+import co.cask.cdap.etl.common.DefaultEmitter;
+
+/**
+ * Key-Value transformation which wraps transformation for each batch plugin to emit stageName along with each record
+ * except for {@link BatchSink}
+ */
+public final class KVTransformation {
+
+  /**
+   * Creates {@link Transformation} which adds stageName to each record being emitted from current stage,
+   * except for {@link BatchSink}. Each stage will strip off stageName, apply original transformation and then again
+   * wrap record with stageName. For {@link co.cask.cdap.etl.api.Joiner} the record is passed with stageName.
+   *
+   * @param stageName      stageName which is emitting the record
+   * @param pluginType     type of the stage
+   * @param isFirstHalf    if first half of reduce type is being executed
+   * @param transformation transformation to be wrapped
+   * @return {@link Transformation} to wrap/unwrap stageName
+   */
+  public Transformation getKVTransformation(String stageName, String pluginType, boolean isFirstHalf,
+                                            Transformation transformation) {
+    if (BatchSink.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      return new KVSinkTransformation<>(transformation);
+    } else if (BatchSource.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      return new KVSourceTransformation<>(stageName, transformation);
+    } else if (Constants.CONNECTOR_TYPE.equalsIgnoreCase(pluginType)) {
+      // TODO Add SparkCompute and SparkSink once CDAP-6065 is merged
```

Comment:
is there a typo in this jira? If its the spark refactor, won't need to do anything for SparkCompute or SparkSink, as that will go through a different code path.